### ISSUE TITLE
company-elm only completes fully qualified names

### ISSUE THREAD

**Issue by mewa** [jcollard/elm-mode#119]
Currently the `company-elm` backend only completes names for fully qualified names even if it's not required.

Here's an example:
<img width="599" alt="zrzut ekranu 2017-10-10 o 15 26 35" src="https://user-images.githubusercontent.com/7347194/31388836-9d9519e8-adcf-11e7-99a1-99591b28658e.png">
<img width="218" alt="zrzut ekranu 2017-10-10 o 15 26 56" src="https://user-images.githubusercontent.com/7347194/31388837-9dcace9e-adcf-11e7-90ab-e8f2cb9e5735.png">

The same happens if you import a module under a different name and try to use it - only the original names (i.e. defined inside the module) will be used.

I have checked `elm-oracle` is working (and you can even check the type of `bod` in the 2nd example with `C-c C-t`) so the issue is with the company backend.

**Comment by purcell** [comment]
I've removed the restriction that completed symbols must contain ".", but there's a little more work to do before this will work properly. Watch this space.

**Comment by mewa** [comment]
I'd be glad to help out. I did start working on it when I submitted the ticket, but it isn't going all that fast since I'm a CL noob

**Comment by purcell** [comment]
Thanks. I'll probably get chance to investigate at the weekend - I'm personally keen to have this working better. I think the solution involves teaching the completion system about the (possibly-aliased) names of the imported modules. This would also help with the slightly broken eldoc support, which causes it to flash up docs for the first _completion_ of the symbol at point, even if the symbol at point is itself invalid.

**Comment by mewa** [comment]
I added alias extraction and storage. It still needs proper wiring to read imports (I was thinking about adding it on save and the `elm-import` command)  as well as changing the completions backend to adapt to these changes. I'll only have a chance to continue in a couple days but I'll need some time to read through the async conses and the like anyway, so if you're willing to build upon it you can have a look at my [fork](https://github.com/jcollard/elm-mode/compare/master...mewa:master).

**Comment by welf** [comment]
I have the same problem. Company autocompletes only if you use names for fully qualified names. Functions, types etc. which I've defined in my file company doesn't complete. 

If I start typing the function's name and run `M-x company-complete` I get the error: 
```
Debugger entered--Lisp error: (error "Company: backend company-elm async timeout with args (candidates updateNam)")
  signal(error ("Company: backend company-elm async timeout with args (candidates updateNam)"))
  error("Company: backend %s async timeout with args %s" company-elm (candidates "updateNam"))
  company--force-sync(company-call-backend-raw (candidates "updateNam") company-elm)
  company-call-backend(candidates "updateNam")
  company--fetch-candidates("updateNam")
  company-calculate-candidates("updateNam")
  company--begin-new()
  company--perform()
  company-auto-begin()
  company-manual-begin()
  company-complete()
  funcall-interactively(company-complete)
  call-interactively(company-complete record nil)
  command-execute(company-complete record)
  helm-M-x(nil #("company-complete" 0 16 (match-part "company-complete")))
  funcall-interactively(helm-M-x nil #("company-complete" 0 16 (match-part "company-complete")))
  call-interactively(helm-M-x nil nil)
  command-execute(helm-M-x)
```
This problem is with Elm files only. In Haskell files company works as expected.

**Comment by purcell** [comment]
@mewa Thanks for sharing your tinkering. I think I have a plan for improving the completion subsystem, which I've laid out in #121.

**Comment by purcell** [comment]
Fixed by #122
